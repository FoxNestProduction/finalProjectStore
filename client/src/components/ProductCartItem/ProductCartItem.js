import React from 'react';
import PropTypes from 'prop-types';
import Card from '@mui/material/Card';
import { CardMedia, IconButton, Typography } from '@mui/material';
import Button from '@mui/material/Button';
import ButtonGroup from '@mui/material/ButtonGroup';
import Box from '@mui/material/Box';
import RemoveRoundedIcon from '@mui/icons-material/RemoveRounded';
import AddRoundedIcon from '@mui/icons-material/AddRounded';
import CloseRoundedIcon from '@mui/icons-material/CloseRounded';
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouterLink } from 'react-router-dom';
import Link from '@mui/material/Link';
import { deleteFromCart, addOneMore, deleteFullProduct } from '../../redux/slices/cartSlice';
import { fixedEncodeURIComponent } from '../../utils/uriEncodeHelpers';
import { totalSumFromCartProduct } from '../Cart/cartFunctions';

import {
  cartIconContainer,
  img,
  textContentBox,
  buttonStyles,
  linkContainer,
  quantityStyle,
  textTitle,
  roundedIcons,
} from './styles';

const ProductCartItem = ({ _id, itemNo, name, cartQuantity, currentPrice, imageUrl }) => {
  const cartProducts = useSelector((state) => state.cart.cart.products);
  const dispatch = useDispatch();
  const index = cartProducts.findIndex(({ product }) => product._id === _id);
  const finallySumOfCartProduct = totalSumFromCartProduct(currentPrice, cartQuantity);
  const handleDeleteOne = () => {
    if (index !== -1) {
      const foundObject = cartProducts[index];
      (() => dispatch(deleteFromCart(foundObject)))();
    }
  };
  const handleAddOne = () => {
    if (index !== -1) {
      const foundObject = cartProducts[index];
      (() => dispatch(addOneMore(foundObject)))();
    }
  };
  const handleDeleteFullProduct = () => {
    if (index !== -1) {
      const foundObject = cartProducts[index];
      dispatch(deleteFullProduct(foundObject));
    }
  };
  return (
    <Card sx={cartIconContainer}>
      <Box sx={linkContainer}>
        <Box
          sx={{
            width: {
              mobile: '40%',
              tablet: '40%',
              desktop: '35%',
            },
          }}
        >
          <Link component={RouterLink} to={`/menu/${fixedEncodeURIComponent(name)}/${itemNo}`}>
            <CardMedia
              component="img"
              sx={img}
              image={imageUrl}
              alt="Live from space album cover"
            />
          </Link>
        </Box>
        <Box sx={textContentBox}>
          <Typography sx={textTitle} variant="body2">
            {name}
          </Typography>
          <Typography variant="h3">
            $
            {finallySumOfCartProduct}
          </Typography>
        </Box>
      </Box>
      <Box
        sx={{
          display: 'flex',
          justifyContent: 'center',
          alignItems: 'center',
          mr: '5px',
          position: {
            mobile: 'relative',
          },
          bottom: {
            mobile: '-12px',
            // tablet: '2px',
            lgTablet: '0px',
            // desktop: '0px',
          },
          right: {
            mobile: '2px',
            // tablet: '4px',
            lgTablet: '0px',
            // desktop: '4px',
          },
        }}
      >
        <ButtonGroup variant="outlined" aria-label="small button group">
          <Button sx={buttonStyles} onClick={handleDeleteOne}>
            <RemoveRoundedIcon sx={roundedIcons} />
          </Button>
          <Typography variant="body2" sx={quantityStyle}>
            {cartQuantity}
          </Typography>
          <Button sx={buttonStyles} onClick={handleAddOne}>
            <AddRoundedIcon sx={roundedIcons} />
          </Button>
        </ButtonGroup>
      </Box>
      <IconButton
        onClick={handleDeleteFullProduct}
        aria-label="CloseRoundedIcon"
        size="small"
        color="secondary"
        sx={{
          p: {
            mobile: '1px',
            lgTablet: '2px',
            desktop: '4px',
          },
          position: 'absolute',
          top: {
            mobile: '3px',
            tablet: '3px',
            lgTablet: '4px',
            desktop: '3px',
          },
          right: {
            mobile: '4px',
            tablet: '4px',
            lgTablet: '4px',
            desktop: '4px',
          },
        }}
      >
        <CloseRoundedIcon
          fontSize="small"
          sx={{
            fontSize: {
              mobile: '16px',
              tablet: '18px',
              lgTablet: '22px',
              desktop: '26px',
            },
          }}
        />
      </IconButton>
    </Card>
  );
};

ProductCartItem.propTypes = {
  currentPrice: PropTypes.number,
  imageUrl: PropTypes.string,
  name: PropTypes.string,
  cartQuantity: PropTypes.number,
  _id: PropTypes.string,
  itemNo: PropTypes.string,
};

ProductCartItem.defaultProps = {
  currentPrice: 12.99,
  imageUrl: './img/salads/3.png',
  name: 'Chicken Hell',
  cartQuantity: 3,
  _id: '',
  itemNo: '',
};

export default ProductCartItem;
