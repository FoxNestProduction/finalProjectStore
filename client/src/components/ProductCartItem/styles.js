export const cartIconContainer = {
  // maxHeight: {
  //   mobile: '120px',
  //   tablet: '135px',
  //   desktop: '143px',
  // },
  maxWidth: {
    mobile: '481px',
    tablet: '993px',
    desktop: '1440px',
  },
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  borderRadius: '18px',
};

export const linkContainer = {
  display: 'flex',
  justifyContent: 'flex-start',
  alignItems: 'center',
  flexWrap: 'nowrap',
  // width: '40%',
  width: '78%', // було 40%
  height: 'auto',
};

export const textContentBox = {
  mx: {
    mobile: '4px',
    tablet: '20px',
    desktop: '30px',
  },
  width: '100%',
};

export const img = {
  minWidth: {
    // mobile: '76px',
    mobile: '52px',
    tablet: '96px',
    desktop: '110px',
  },
  // maxWidth: {
  //   mobile: '106px',
  //   tablet: '127px',
  // },
  // maxWidth: '100%',
  height: 'auto',
  objectFit: 'cover',
  // borderRadius: '8px',
};

export const textContent = {

};

export const textTitle = {
  fontSize: {
    mobile: '14px',
    tablet: '18px',
    desktop: '20px',
  },
  lineHeight: {
    mobile: '15px',
    tablet: '23px',
    desktop: '26px',
  },
  weight: 600,
  display: 'flex',
  width: {
    mobile: '100px',
    // tablet: '160px',
    tablet: '100%',
  },
  flexWrap: 'wrap',
  // width: '100%',
  // overflow: 'hidden',
  // whiteSpace: 'nowrap',
  textDecoration: 'none',
};

export const buttonStyles = {
  padding: {
    // mobile: '2px 7px',
    mobile: '0px 5px',
    tablet: '0 8px',
    desktop: '1px 11px',
  },
  borderRadius: '8px',
  minWidth: '28px',
};

export const quantityStyle = {
  border: '1px solid rgba(108, 095, 188, 0.5)',
  // px: '10px',
  px: {
    mobile: '0',
    tablet: '10px',
  },
  fontSize: {
    mobile: '14px',
    tablet: '18px',
    desktop: '20px',
  },
  width: '2rem',
  textAlign: 'center',
};

// export const buttonStyles = {
//   padding: {
//     lgTablet: '0 5px',
//   },
//   borderRadius: '8px',
//   minWidth: '32px',
// };

export const roundedIcons = {
  fontSize: {
    mobile: '1.2rem',
    tablet: '1.5rem',
  },
};

// export const quantityStyle = {
//   border: '1px solid rgba(108, 095, 188, 0.5)',
//   pt: '1px',
//   fontWeight: 'fontWeightMedium',
//   fontSize: {
//     lgTablet: '14px',
//   },
//   width: '2rem',
//   textAlign: 'center',
// };
