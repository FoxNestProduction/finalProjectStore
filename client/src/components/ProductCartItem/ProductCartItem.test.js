import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, screen } from '@testing-library/react';
import ProductCartItem from './ProductCartItem';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

const mockDispatch = jest.spyOn(require('react-redux'), 'useDispatch');

describe('ProductCartItem Component', () => {

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should renders ProductCartItem component', () => {
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    const { asFragment } = render(
      <MemoryRouter>
          <ProductCartItem
            name="Product"
            itemNo="123"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('should delete full product when click dell', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    render(
      <MemoryRouter>
          <ProductCartItem
            name="Product"
            itemNo="123"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    const deleteButtons = screen.getAllByLabelText('CloseRoundedIcon');
    fireEvent.click(deleteButtons[0]);
    expect(dispatch).toHaveBeenCalled();
  });

  test('should delete one product when click dell', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    render(
      <MemoryRouter>
          <ProductCartItem
            name="Product"
            itemNo="123"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    const deleteButtons = screen.getAllByTestId('RemoveRoundedIcon');
    fireEvent.click(deleteButtons[0]);
    expect(dispatch).toHaveBeenCalled();
  });

  test('should added one product when click dell Auth', () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce([]);
    jest.spyOn(require('react-redux'), 'useSelector').mockReturnValueOnce(true);

    render(
      <MemoryRouter>
          <ProductCartItem
            name="Product"
            itemNo="123"
            currentPrice={10.99}
            imageUrl="image.jpg"
            cartQuantity={1}
            _id="123"
          />
      </MemoryRouter>,
    );

    const addButtons = screen.getAllByTestId('AddRoundedIcon');
    fireEvent.click(addButtons[0]);
    expect(dispatch).toHaveBeenCalled();
  });
});