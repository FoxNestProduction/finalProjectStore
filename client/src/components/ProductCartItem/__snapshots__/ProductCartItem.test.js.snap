// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCartItem Component should renders ProductCartItem component 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-3mr10w-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-wflwa4"
    >
      <div
        class="MuiBox-root css-11f0zxs"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1xylxj1-MuiTypography-root-MuiLink-root"
          href="/menu/product/123"
        >
          <img
            alt="Live from space album cover"
            class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1bcnftm-MuiCardMedia-root"
            src="image.jpg"
          />
        </a>
      </div>
      <div
        class="MuiBox-root css-euwkvv"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1xbu4ut-MuiTypography-root"
        >
          Product
        </p>
        <h3
          class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
        >
          $10.99
        </h3>
      </div>
    </div>
    <div
      class="MuiBox-root css-15ju9rx"
    >
      <div
        aria-label="small button group"
        class="MuiButtonGroup-root MuiButtonGroup-outlined css-edsu1d-MuiButtonGroup-root"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-firstButton css-nf8asi-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1jth6sh-MuiSvgIcon-root"
            data-testid="RemoveRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 13H6c-.55 0-1-.45-1-1s.45-1 1-1h12c.55 0 1 .45 1 1s-.45 1-1 1z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <p
          class="MuiTypography-root MuiTypography-body2 css-kseybu-MuiTypography-root"
        >
          1
        </p>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlinedPrimary MuiButtonGroup-lastButton css-nf8asi-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1jth6sh-MuiSvgIcon-root"
            data-testid="AddRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <button
      aria-label="CloseRoundedIcon"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall css-15ufy6i-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-1btoxe8-MuiSvgIcon-root"
        data-testid="CloseRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18.3 5.71a.9959.9959 0 0 0-1.41 0L12 10.59 7.11 5.7a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
